### PR TITLE
Refactor the migration code for the lucene module

### LINKED ISSUE #13488: The migration script of the Lucene Module  does not consider compatibility with other databases
`session.Store.Configuration.SqlDialec` is only for SqlServer  , we should use `SchemaBuilder.Dialec` to instead

https://github.com/OrchardCMS/OrchardCore/blob/afc99f106b97b50439f73636bd9c124923d41311/src/OrchardCore.Modules/OrchardCore.Search.Lucene/Migrations.cs#L169-L200

```
An error occurred while updating Lucene indices settings and queries Npgsql.PostgresException (0x80004005): 42601: syntax error at or near "["

POSITION: 198
   at Npgsql.Internal.NpgsqlConnector.<ReadMessage>g__ReadMessageLong|221_0(NpgsqlConnector connector, Boolean async, DataRowLoadingMode dataRowLoadingMode, Boolean readingNotifications, Boolean isReadingPrependedMessage)
   at Npgsql.NpgsqlDataReader.NextResult(Boolean async, Boolean isConsuming, CancellationToken cancellationToken)
   at Npgsql.NpgsqlDataReader.NextResult(Boolean async, Boolean isConsuming, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteReader(CommandBehavior behavior, Boolean async, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteReader(CommandBehavior behavior, Boolean async, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteNonQuery(Boolean async, CancellationToken cancellationToken)
   at Dapper.SqlMapper.ExecuteImplAsync(IDbConnection cnn, CommandDefinition command, Object param) in /_/Dapper/SqlMapper.Async.cs:line 646
   at OrchardCore.Search.Lucene.Migrations.<>c.<<Create>b__2_0>d.MoveNext()
  Exception data:
    Severity: ERROR
    SqlState: 42601
    MessageText: syntax error at or near "["
    Position: 198
    File: scan.l
    Line: 1133
    Routine: scanner_yyerror    at Npgsql.Internal.NpgsqlConnector.<ReadMessage>g__ReadMessageLong|221_0(NpgsqlConnector connector, Boolean async, DataRowLoadingMode dataRowLoadingMode, Boolean readingNotifications, Boolean isReadingPrependedMessage)
   at Npgsql.NpgsqlDataReader.NextResult(Boolean async, Boolean isConsuming, CancellationToken cancellationToken)
   at Npgsql.NpgsqlDataReader.NextResult(Boolean async, Boolean isConsuming, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteReader(CommandBehavior behavior, Boolean async, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteReader(CommandBehavior behavior, Boolean async, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteNonQuery(Boolean async, CancellationToken cancellationToken)
   at Dapper.SqlMapper.ExecuteImplAsync(IDbConnection cnn, CommandDefinition command, Object param) in /_/Dapper/SqlMapper.Async.cs:line 646
   at OrchardCore.Search.Lucene.Migrations.<>c.<<Create>b__2_0>d.MoveNext()
```

### PR DESCRIPTION
**Pull request by hyzx86**
Fix #13488

### CODE REVIEW

**Comment by hishamco** [comment]
I thought there was a bug when I read the title :) may be we could use clean or refactor instead

File: src/OrchardCore.Modules/OrchardCore.Search.Lucene/Migrations.cs [hunk 1]
```diff
@@ -166,36 +166,39 @@ public int Create()
 
                     using (var transaction = connection.BeginTransaction(session.Store.Configuration.IsolationLevel))
                     {
-                        var dialect = session.Store.Configuration.SqlDialect;
+                        var dialect = SchemaBuilder.Dialect;
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Search.Lucene/Migrations.cs:169]: Here you are in the delegate of a deferred task, not good to keep a reference on the SchemaBuilder of the Migrations.

Anyway in YesSql `SchemaBuilder.Dialect` is initialized with `configuration.SqlDialect`.

Maybe the issue is only due to the missing usage of the below `dialect.QuoteForColumnName()`.

Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Search.Lucene/Migrations.cs]: updated

File: src/OrchardCore.Modules/OrchardCore.Search.Lucene/Migrations.cs [hunk 2]
```diff
@@ -157,7 +157,7 @@ public int Create()
                 var logger = scope.ServiceProvider.GetService<ILogger<Migrations>>();
                 var tablePrefix = session.Store.Configuration.TablePrefix;
                 var documentTableName = session.Store.Configuration.TableNameConvention.GetDocumentTable();
-
+
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Search.Lucene/Migrations.cs:160]: Remove extra spaces, under Visual Studio you can do ctrl+K+D to format the file.

File: src/OrchardCore.Modules/OrchardCore.Search.Lucene/Migrations.cs [hunk 3]
```diff
@@ -174,28 +174,31 @@ public int Create()
                             {
                                 logger.LogDebug("Updating Lucene indices settings and queries");
                             }
+                            var quoteTableName = dialect.QuoteForTableName(table, session.Store.Configuration.Schema);
+                            var quoteContentColName = dialect.QuoteForColumnName("Content");
+                            var quoteTypeColName = dialect.QuoteForColumnName("Type");
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Search.Lucene/Migrations.cs:179]: I suggest `var quotedTableName`, `quotedContentColumnName` and `quotedTypeColumnName`.

**Comment by jtkech** [comment]
I let you retry it.

Let me know if it still works as expected so that tomorrow I will approve.

File: src/OrchardCore.Modules/OrchardCore.Search.Lucene/Migrations.cs (hunk 2 above)
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Search.Lucene/Migrations.cs:160]:
```suggestion

```

File: src/OrchardCore.Modules/OrchardCore.Search.Lucene/Migrations.cs [hunk 4]
```diff
@@ -174,28 +174,31 @@ public int Create()
...
                             }
+                            var quotedTableName = dialect.QuoteForTableName(table, session.Store.Configuration.Schema);
+                            var quotedContentColumnName = dialect.QuoteForColumnName("Content");
+                            var quotedTypeColumnName = dialect.QuoteForColumnName("Type");
 
-                            var updateCmd = $"UPDATE {dialect.QuoteForTableName(table, session.Store.Configuration.Schema)} SET Content = REPLACE(content, '\"$type\":\"OrchardCore.Lucene.LuceneQuery, OrchardCore.Lucene\"', '\"$type\":\"OrchardCore.Search.Lucene.LuceneQuery, OrchardCore.Search.Lucene\"') WHERE [Type] = 'OrchardCore.Queries.Services.QueriesDocument, OrchardCore.Queries'";
+                            var updateCmd = $"UPDATE {quotedTableName} SET {quotedContentColumnName} = REPLACE({quotedContentColumnName}, '\"$type\":\"OrchardCore.Lucene.LuceneQuery, OrchardCore.Lucene\"', '\"$type\":\"OrchardCore.Search.Lucene.LuceneQuery, OrchardCore.Search.Lucene\"') WHERE {quotedTypeColumnName}  = 'OrchardCore.Queries.Services.QueriesDocument, OrchardCore.Queries'";
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Search.Lucene/Migrations.cs:181]: Why did `Content` change to quotedContentColumnName?

**Review by hishamco — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Modules/OrchardCore.Search.Lucene/Migrations.cs (hunk 4 above)
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Search.Lucene/Migrations.cs:180]: It looks like you have to use double quotes around column names in pgsql


image

Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Search.Lucene/Migrations.cs:180]: you can try this script 

``` sql
UPDATE "Document"
SET Content = REPLACE(Content, '\"$type\":\"OrchardCore.Lucene.LuceneQuery, OrchardCore.Lucene\"', '\"$type\":\"OrchardCore.Search.Lucene.LuceneQuery, OrchardCore.Search.Lucene\"') 
WHERE Type  = 'OrchardCore.Queries.Services.QueriesDocument, OrchardCore.Queries'
```

**Comment by hyzx86** [comment]
> I let you retry it.
> 
> Let me know if it still works as expected so that tomorrow I will approve.
@jtkech I've tested it on postgres and nothing is wrong

**Review by jtkech — APPROVED** [review]

**Review by sebastienros — APPROVED** [review]

**Comment by sebastienros** [comment]
This should have been part of 1.6

**Comment by sebastienros** [comment]
What will happen for users that have had this Create fail, and try to run this updated version. Will 'Create' be tried again?

**Comment by hyzx86** [comment]
> What will happen for users that have had this Create fail, and try to run this updated version. Will 'Create' be tried again?

I don't remember, I was testing the deployment under different databases, I tried Mysql and pgsql, both of them had this error, and then I looked at the source code, if the user is installing OC for the first time, even if it runs successfully, nothing changes

**Comment by Skrypt** [comment]
Why is it not merged?

**Comment by hyzx86** [comment]
Hi @hishamco ,Is there any problem to deal with this PR?

**Comment by hishamco** [comment]
Seems @jtkech merge this before I reply :)

**Pull request merged**